File: services/backend/branding_config.py
"""
Branding Configuration Loader for Backend

This module loads branding configuration from the root branding.yaml file
and makes it available to the Flask backend.
"""

import os
import yaml
from typing import Dict, Any, Optional
from pathlib import Path


class BrandingConfig:
    """Loads and manages branding configuration."""

    _instance: Optional['BrandingConfig'] = None
    _config: Dict[str, Any] = {}

    def __new__(cls):
        """Singleton pattern to ensure single config instance."""
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._load_config()
        return cls._instance

    def _load_config(self) -> None:
        """Load branding configuration from branding.yaml."""
        # Try to find branding.yaml in project root
        current_dir = Path(__file__).resolve()
        project_root = current_dir.parent.parent.parent
        branding_file = project_root / "branding.yaml"

        if not branding_file.exists():
            print(f"Warning: branding.yaml not found at {branding_file}")
            print("Using default configuration")
            self._config = self._get_default_config()
            return

        try:
            with open(branding_file, 'r', encoding='utf-8') as f:
                self._config = yaml.safe_load(f)
            print(f"Loaded branding configuration from {branding_file}")
        except Exception as e:
            print(f"Error loading branding.yaml: {e}")
            print("Using default configuration")
            self._config = self._get_default_config()

    def _get_default_config(self) -> Dict[str, Any]:
        """Return default configuration if branding.yaml is not found."""
        return {
            'assistant': {
                'name': 'Assistant',
                'role': 'AI assistant',
                'tagline': 'Your AI Companion',
                'description': 'An AI assistant ready to help.',
                'credentials': 'AI assistant'
            },
            'ui': {
                'page_title': '{assistant_name} - AI Assistant',
                'header_text': '{assistant_name}',
                'welcome_message': 'Welcome to {assistant_name}',
                'chat': {
                    'assistant_label': '{assistant_name}',
                    'user_label': 'You',
                    'status': {
                        'thinking': '{assistant_name} is thinking...',
                        'speaking': '{assistant_name} is speaking...',
                        'typing': '{assistant_name} is typing...',
                        'listening': 'Listening...',
                        'processing': 'Processing your message...'
                    }
                }
            },
            'visual': {
                'logo': {
                    'path': '',
                    'alt_text': '{assistant_name} Logo',
                    'width': '150px',
                    'height': '50px'
                },
                'colors': {
                    'primary': '#0066cc',
                    'secondary': '#667eea',
                    'accent': '#764ba2',
                    'gradient': 'linear-gradient(135deg, #667eea 0%, #764ba2 100%)',
                    'text_primary': '#333333',
                    'text_secondary': '#666666',
                    'text_light': '#ffffff'
                },
                'favicon': ''
            },
            'service': {
                'backend_name': 'AI Backend',
                'ai_backend_name': 'AI Service',
                'database_name': 'assistant',
                'session_timeout': 30
            },
            'features': {
                'voice_chat_enabled': True,
                'risk_assessment_enabled': True,
                'transcripts_enabled': True,
                'admin_dashboard_enabled': True
            }
        }

    def _interpolate(self, text: str) -> str:
        """Replace placeholders in text with actual values."""
        if not isinstance(text, str):
            return text

        replacements = {
            '{assistant_name}': self.get('assistant.name', 'Assistant'),
            '{role}': self.get('assistant.role', 'AI assistant'),
            '{credentials}': self.get('assistant.credentials', 'AI assistant'),
            '{tagline}': self.get('assistant.tagline', 'Your AI Companion')
        }

        result = text
        for placeholder, value in replacements.items():
            result = result.replace(placeholder, value)

        return result

    def get(self, key: str, default: Any = None) -> Any:
        """
        Get a configuration value using dot notation.

        Args:
            key: Configuration key in dot notation (e.g., 'assistant.name')
            default: Default value if key is not found

        Returns:
            Configuration value with placeholders replaced
        """
        keys = key.split('.')
        value = self._config

        for k in keys:
            if isinstance(value, dict) and k in value:
                value = value[k]
            else:
                return default

        # Interpolate placeholders if value is a string
        if isinstance(value, str):
            return self._interpolate(value)

        return value

    def get_all(self) -> Dict[str, Any]:
        """Get the entire configuration dictionary."""
        return self._config

    def get_frontend_config(self) -> Dict[str, Any]:
        """Get configuration formatted for frontend consumption."""
        return {
            'assistant': {
                'name': self.get('assistant.name'),
                'tagline': self.get('assistant.tagline'),
                'description': self.get('assistant.description')
            },
            'ui': {
                'pageTitle': self.get('ui.page_title'),
                'headerText': self.get('ui.header_text'),
                'welcomeMessage': self.get('ui.welcome_message'),
                'chat': {
                    'assistantLabel': self.get('ui.chat.assistant_label'),
                    'userLabel': self.get('ui.chat.user_label'),
                    'status': {
                        'thinking': self.get('ui.chat.status.thinking'),
                        'speaking': self.get('ui.chat.status.speaking'),
                        'typing': self.get('ui.chat.status.typing'),
                        'listening': self.get('ui.chat.status.listening'),
                        'processing': self.get('ui.chat.status.processing')
                    }
                }
            },
            'visual': {
                'logo': {
                    'path': self.get('visual.logo.path'),
                    'altText': self.get('visual.logo.alt_text'),
                    'width': self.get('visual.logo.width'),
                    'height': self.get('visual.logo.height')
                },
                'colors': {
                    'primary': self.get('visual.colors.primary'),
                    'secondary': self.get('visual.colors.secondary'),
                    'accent': self.get('visual.colors.accent'),
                    'gradient': self.get('visual.colors.gradient'),
                    'textPrimary': self.get('visual.colors.text_primary'),
                    'textSecondary': self.get('visual.colors.text_secondary'),
                    'textLight': self.get('visual.colors.text_light')
                },
                'favicon': self.get('visual.favicon')
            },
            'features': {
                'voiceChatEnabled': self.get('features.voice_chat_enabled', True),
                'riskAssessmentEnabled': self.get('features.risk_assessment_enabled', True),
                'transcriptsEnabled': self.get('features.transcripts_enabled', True),
                'adminDashboardEnabled': self.get('features.admin_dashboard_enabled', True)
            }
        }

    def reload(self) -> None:
        """Reload configuration from file."""
        self._load_config()


# Global instance
_branding_config = None


def get_branding_config() -> BrandingConfig:
    """Get the global branding configuration instance."""
    global _branding_config
    if _branding_config is None:
        _branding_config = BrandingConfig()
    return _branding_config

File: services/frontend/static/js/branding-config.js
/**
 * Branding Configuration Manager
 *
 * Fetches and manages branding configuration from the backend.
 * Provides a centralized way to access branding settings throughout the frontend.
 */

class BrandingConfig {
    constructor() {
        this.config = null;
        this.loading = false;
        this.loadPromise = null;
    }

    /**
     * Load branding configuration from the backend API
     * @returns {Promise<Object>} The branding configuration
     */
    async load() {
        // Return existing load promise if already loading
        if (this.loading && this.loadPromise) {
            return this.loadPromise;
        }

        // Return cached config if already loaded
        if (this.config) {
            return Promise.resolve(this.config);
        }

        this.loading = true;
        this.loadPromise = this._fetchConfig();

        try {
            this.config = await this.loadPromise;
            this.loading = false;
            return this.config;
        } catch (error) {
            this.loading = false;
            this.loadPromise = null;
            throw error;
        }
    }

    /**
     * Internal method to fetch configuration from backend
     * @private
     */
    async _fetchConfig() {
        try {
            const response = await fetch('http://localhost:5000/api/branding', {
                method: 'GET',
                credentials: 'include',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            if (!response.ok) {
                throw new Error(`Failed to fetch branding config: ${response.statusText}`);
            }

            const data = await response.json();

            if (data.success && data.data) {
                return data.data;
            } else {
                throw new Error('Invalid branding configuration response');
            }
        } catch (error) {
            console.error('Error loading branding configuration:', error);
            // Return default configuration on error
            return this._getDefaultConfig();
        }
    }

    /**
     * Get default configuration (fallback)
     * @private
     */
    _getDefaultConfig() {
        return {
            assistant: {
                name: 'Assistant',
                tagline: 'Your AI Companion',
                description: 'An AI assistant ready to help.'
            },
            ui: {
                pageTitle: 'AI Assistant',
                headerText: 'Assistant',
                welcomeMessage: 'Welcome to Assistant',
                chat: {
                    assistantLabel: 'Assistant',
                    userLabel: 'You',
                    status: {
                        thinking: 'Assistant is thinking...',
                        speaking: 'Assistant is speaking...',
                        typing: 'Assistant is typing...',
                        listening: 'Listening...',
                        processing: 'Processing your message...'
                    }
                }
            },
            visual: {
                logo: {
                    path: '',
                    altText: 'Assistant Logo',
                    width: '150px',
                    height: '50px'
                },
                colors: {
                    primary: '#0066cc',
                    secondary: '#667eea',
                    accent: '#764ba2',
                    gradient: 'linear-gradient(135deg, #667eea 0%, #764ba2 100%)',
                    textPrimary: '#333333',
                    textSecondary: '#666666',
                    textLight: '#ffffff'
                },
                favicon: ''
            },
            features: {
                voiceChatEnabled: true,
                riskAssessmentEnabled: true,
                transcriptsEnabled: true,
                adminDashboardEnabled: true
            }
        };
    }

    /**
     * Get a configuration value using dot notation
     * @param {string} path - Path to config value (e.g., 'assistant.name')
     * @param {*} defaultValue - Default value if path not found
     * @returns {*} Configuration value
     */
    get(path, defaultValue = null) {
        if (!this.config) {
            console.warn('Branding config not loaded yet');
            return defaultValue;
        }

        const keys = path.split('.');
        let value = this.config;

        for (const key of keys) {
            if (value && typeof value === 'object' && key in value) {
                value = value[key];
            } else {
                return defaultValue;
            }
        }

        return value;
    }

    /**
     * Get assistant name
     * @returns {string}
     */
    getAssistantName() {
        return this.get('assistant.name', 'Assistant');
    }

    /**
     * Get assistant tagline
     * @returns {string}
     */
    getTagline() {
        return this.get('assistant.tagline', 'Your AI Companion');
    }

    /**
     * Get page title
     * @returns {string}
     */
    getPageTitle() {
        return this.get('ui.pageTitle', 'AI Assistant');
    }

    /**
     * Get chat status message
     * @param {string} type - Status type (thinking, speaking, typing, etc.)
     * @returns {string}
     */
    getStatusMessage(type) {
        return this.get(`ui.chat.status.${type}`, 'Processing...');
    }

    /**
     * Get assistant label for chat messages
     * @returns {string}
     */
    getAssistantLabel() {
        return this.get('ui.chat.assistantLabel', 'Assistant');
    }

    /**
     * Get user label for chat messages
     * @returns {string}
     */
    getUserLabel() {
        return this.get('ui.chat.userLabel', 'You');
    }

    /**
     * Get primary color
     * @returns {string}
     */
    getPrimaryColor() {
        return this.get('visual.colors.primary', '#0066cc');
    }

    /**
     * Get gradient
     * @returns {string}
     */
    getGradient() {
        return this.get('visual.colors.gradient', 'linear-gradient(135deg, #667eea 0%, #764ba2 100%)');
    }

    /**
     * Apply branding to the page
     * Updates page title, favicon, CSS variables, etc.
     */
    applyBranding() {
        if (!this.config) {
            console.warn('Cannot apply branding - config not loaded');
            return;
        }

        // Update page title
        const pageTitle = this.getPageTitle();
        if (pageTitle) {
            document.title = pageTitle;
        }

        // Update favicon if specified
        const faviconPath = this.get('visual.favicon');
        if (faviconPath) {
            let link = document.querySelector("link[rel~='icon']");
            if (!link) {
                link = document.createElement('link');
                link.rel = 'icon';
                document.head.appendChild(link);
            }
            link.href = faviconPath;
        }

        // Apply CSS variables for colors
        const colors = this.get('visual.colors', {});
        const root = document.documentElement;

        if (colors.primary) root.style.setProperty('--brand-primary', colors.primary);
        if (colors.secondary) root.style.setProperty('--brand-secondary', colors.secondary);
        if (colors.accent) root.style.setProperty('--brand-accent', colors.accent);
        if (colors.gradient) root.style.setProperty('--brand-gradient', colors.gradient);
        if (colors.textPrimary) root.style.setProperty('--text-primary', colors.textPrimary);
        if (colors.textSecondary) root.style.setProperty('--text-secondary', colors.textSecondary);
        if (colors.textLight) root.style.setProperty('--text-light', colors.textLight);

        console.log('Branding applied successfully');
    }

    /**
     * Reload configuration from backend
     * @returns {Promise<Object>}
     */
    async reload() {
        this.config = null;
        this.loading = false;
        this.loadPromise = null;
        return await this.load();
    }
}

// Create global instance
const brandingConfig = new BrandingConfig();

// Auto-load configuration when script is included
if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', async () => {
        try {
            await brandingConfig.load();
            brandingConfig.applyBranding();
        } catch (error) {
            console.error('Failed to load branding configuration:', error);
        }
    });
} else {
    // DOM already loaded
    brandingConfig.load().then(() => {
        brandingConfig.applyBranding();
    }).catch(error => {
        console.error('Failed to load branding configuration:', error);
    });
}

// Export for use in other scripts
window.brandingConfig = brandingConfig;
